File: topcoder/tco/13/1a/DirectionBoard.cpp
#include <algorithm>
#include <string>
#include <vector>

using namespace std;

#define INF 123456789
#define N 5005
#define M 30005

#define in(i,j) 2 * ( i * y + j )
#define out(i,j) 2 * ( i * y + j ) + 1

char dc[] = { 'D', 'R', 'U', 'L' };
int dx[] = { 1, 0, -1, 0 }, dy[] = { 0, 1, 0, -1 };
int n, m, a[M], b[M], cap[M], cost[M], flow[M], dist[N], prev[N];

inline void add_edge( int u, int v, int _cap, int _cost )
{
	a[m] = u, b[m] = v, cap[m] = _cap, cost[m] = _cost, m++;
	a[m] = v, b[m] = u, cap[m] = 0, cost[m] = -_cost, m++;
}

void min_cost_max_flow( int s, int t, int & mc, int & mf )
{
	mc = mf = 0;
	for( int i = 0; i < m; i++ ) flow[i] = 0;
	while(1)
	{
		for( int i = 0; i < n; i++ ) dist[i] = INF, prev[i] = -1;
		dist[s] = 0;
		for( int i = 0; i < n; i++ )
		{
			bool change = false;
			for( int j = 0; j < m; j++ )
			if( flow[j] < cap[j] && dist[a[j]] + cost[j] < dist[b[j]] )
			{
				dist[b[j]] = dist[a[j]] + cost[j];
				prev[b[j]] = j;
				change = true;
			}
			if( !change ) break;
		}

		if( prev[t] != -1 )
		{
			int f = INF;
			for( int j = prev[t]; j != -1; j = prev[a[j]] ) f = min( f, cap[j] - flow[j] );
			for( int j = prev[t]; j != -1; j = prev[a[j]] ) flow[j] += f, flow[j^1] -= f;
			mf += f, mc += f * dist[t];
		}
		else break;
	}
}

struct DirectionBoard
{
	int getMinimum( vector<string> g )
	{
		int x = g.size(), y = g[0].size(), mc, mf;
		n = 2 * x * y + 2;
		for( int i = 0; i < x; i++ ) for( int j = 0; j < y; j++ ) 
		{
			for( int d = 0; d < 4; d++ )
			{
				int ii = ( i + dx[d] + x ) % x, jj = ( j + dy[d] + y ) % y;
				add_edge( in( i, j ), out( ii, jj ), 1, g[i][j] != dc[d] );
			}
			add_edge( n - 2, in( i, j ), 1, 0 );
			add_edge( out( i, j ), n - 1, 1, 0 );
		}
		min_cost_max_flow( n - 2, n - 1, mc, mf );
		return mc;
	}
};

File: topcoder/tco/13/1a/DirectionBoard.java
import java.math.*;
import java.util.*;

public class DirectionBoard
{
	static char[] dc = {'D', 'R', 'U', 'L'};
	static int[] dx = {1, 0, -1, 0}, dy = {0, 1, 0, -1};
	static int INF = 12345;

	int x, y;
	int in(int i, int j) { return 2 * (i * y + j); }
	int out(int i, int j) { return 2 * (i * y + j) + 1; }

	public int getMinimum(String[] b)
	{
		x = b.length;
		y = b[0].length();
		n = 2 * x * y + 2;
		g = (ArrayList<Edge>[])(new ArrayList[n]);
		for(int i = 0; i < n; i++) g[i] = new ArrayList<Edge>();
		for(int i = 0; i < x; i++) for(int j = 0; j < y; j++)
		{
			addEdge(n - 2, in(i, j), 1, 0);
			addEdge(out(i, j), n - 1, 1, 0);
			for(int d = 0; d < 4; d++)
				addEdge(
					in(i, j), out((i + dx[d] + x) % x, (j + dy[d] + y) % y),
					1, b[i].charAt(j) == dc[d] ? 0 : 1
				);
		}
		return minCostMaxFlow(n - 2, n - 1)[0];
	}

	ArrayList<Edge>[] g;
	int n;

	class Edge
	{
		int j, ca, co, f;
		Edge r;

		Edge(int j, int ca, int co)
		{
			this.j = j;
			this.ca = ca;
			this.co = co;
		}
	}

	void addEdge(int i, int j, int ca, int co)
	{
		Edge e = new Edge(j, ca, co), f = new Edge(i, 0, -co);
		e.r = f;
		f.r = e;
		g[i].add(e);
		g[j].add(f);
	}

	int[] minCostMaxFlow(int s, int t)
	{
		int mc = 0, mf = 0, dist[] = new int[n];
		Edge[] pre = new Edge[n];
		while(true)
		{
			Arrays.fill(dist, INF);
			Arrays.fill(pre, null);
			dist[s] = 0;
			for(int q = 0; q < n; q++)
			{
				boolean change = false;
				for(int i = 0; i < n; i++) for(Edge e : g[i])
				if(e.f < e.ca && dist[i] + e.co < dist[e.j])
				{
					dist[e.j] = dist[i] + e.co;
					pre[e.j] = e;
					change = true;
				}
				if(!change) break;
			}
			if(pre[t] == null) break;
			int f = INF;
			for(Edge e = pre[t]; e != null; e = pre[e.r.j])
				f = Math.min(f, e.ca - e.f);
			for(Edge e = pre[t]; e != null; e = pre[e.r.j])
			{
				e.f += f;
				e.r.f -= f;
			}
			mf += f;
			mc += f * dist[t];
		}
		return new int[]{mc, mf};
	}
}
